storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3MultiPartOutputStream.java

```diff
@@ -86,54 +85,72 @@ public void write(final byte[] b, final int off, final int len) throws IOExcepti
         if (b.length == 0) {
             return;
         }
-        final ByteBuffer source = ByteBuffer.wrap(b, off, len);
-        while (source.hasRemaining()) {
-            final int transferred = Math.min(partBuffer.remaining(), source.remaining());
-            final int offset = source.arrayOffset() + source.position();
-            // TODO: get rid of this array copying
-            partBuffer.put(source.array(), offset, transferred);
-            source.position(source.position() + transferred);
-            if (!partBuffer.hasRemaining()) {
-                flushBuffer(0, partSize);
+        try {
+            final ByteBuffer source = ByteBuffer.wrap(b, off, len);
+            while (source.hasRemaining()) {
+                final int transferred = Math.min(partBuffer.remaining(), source.remaining());
+                final int offset = source.arrayOffset() + source.position();
+                // TODO: get rid of this array copying
+                partBuffer.put(source.array(), offset, transferred);
+                source.position(source.position() + transferred);
+                if (!partBuffer.hasRemaining()) {
+                    flushBuffer(0, partSize);
+                }
             }
+        } catch (final Exception e) {
+            log.error("Failed to write to stream on upload {}, aborting transaction", uploadId, e);
+            abortUpload();
+            throw new IOException(e);
         }
     }
 
     @Override
     public void close() throws IOException {
-        if (partBuffer.position() > 0) {
-            flushBuffer(partBuffer.arrayOffset(), partBuffer.position());
-        }
-        if (Objects.nonNull(uploadId)) {
+        if (closed) {
+            log.warn("Already closed");
```

Comment:
I think we shouldn't even have this logging: this is not a condition we should be warned about and we can't do nothing about it. IMO just `return` is totally sufficient here.